File: archive/2026-04-05/ch06_tree/t17_build_tree.py
"""LeetCode 106. Construct Binary Tree from Inorder and Postorder Traversal
https://leetcode.cn/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
"""

from typing import List, Optional, Dict

from tree_node import TreeNode


#             |idx|
# inorder = [9,3,15,20,7]
#               |       |idx
# postorder = [9,15,7,20,3]
class Solution:
    def buildTree(
            self, inorder: List[int], postorder: List[int]
    ) -> Optional[TreeNode]:
        dic: Dict[int, int] = {}
        for idx, num in enumerate(inorder):
            dic[num] = idx

        def build(
                inorder: List[int], in_start: int, in_end: int,
                postorder: List[int], post_start: int, post_end: int
        ) -> Optional[TreeNode]:
            if in_end == in_start:
                return None
            elif in_end == in_start + 1:
                return TreeNode(val=postorder[post_end - 1])

            root_val = postorder[post_end - 1]
            root_idx = dic[root_val]
            root = TreeNode(val=root_val)

            root.left = build(
                inorder, in_start, root_idx,
                postorder, post_start, post_start + root_idx - in_start
            )
            root.right = build(
                inorder, root_idx + 1, in_end,
                postorder, post_start + root_idx - in_start, post_end - 1
            )
            return root

        return build(inorder, 0, len(inorder), postorder, 0, len(postorder))


if __name__ == "__main__":
    solution = Solution()
    root1 = solution.buildTree([9, 3, 15, 20, 7], [9, 15, 7, 20, 3])
    assert root1.val == 3
    assert root1.left.val == 9
    assert root1.right.val == 20
    assert root1.right.left.val == 15
    assert root1.right.right.val == 7

    root2 = solution.buildTree([-1], [-1])
    assert root2 is not None and root2.val == -1
    print("All tests passed.")

File: archive/2024-12-06/tree/T17_buildTree.java
package org.review2024winter.tree;

import java.util.HashMap;
import java.util.Map;

public class T17_buildTree {
    Map<Integer, Integer> map = new HashMap<>();

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        for (int i = 0; i < postorder.length; i++) {
            map.put(inorder[i], i);
        }
        int n = inorder.length;
        return build(inorder, 0, n, postorder, 0, n);
    }

    public TreeNode build(int[] inorder, int inBegin, int inEnd, int[] postorder, int postBegin, int postEnd) {
        if (inEnd - inBegin == 0) {
            return null;
        } else if (inEnd - inBegin == 1) {
            return new TreeNode(inorder[inBegin]);
        }
        int midVal = postorder[postEnd - 1];
        int midIdx = map.get(midVal);
        TreeNode midRoot = new TreeNode(midVal);
        midRoot.left = build(inorder, inBegin, midIdx, postorder, postBegin, postBegin + midIdx - inBegin);
        midRoot.right = build(inorder, midIdx + 1, inEnd, postorder, postBegin + midIdx - inBegin, postEnd - 1);
        return midRoot;
    }
}
